File: examples/python/obsidian_memory_client.py
#!/usr/bin/env python3
"""Obsidian-Memory Python SDK Example.

A simple client library for interacting with the Obsidian-Memory API.

Usage:
    from obsidian_memory_client import ObsidianMemoryClient

    client = ObsidianMemoryClient(
        base_url="http://localhost:8000",
        auth_token="your-token"  # Optional if auth is disabled
    )

    # List notes
    notes = client.list_notes(limit=10)

    # Search notes
    results = client.search("machine learning", project="research")

    # Create a note
    note = client.create_note(
        vault_name="my-vault",
        title="New Note",
        content="# My Note\\n\\nContent here..."
    )
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime
from typing import Any

import httpx


@dataclass
class Note:
    """Represents a note in the knowledge base."""

    id: int
    vault_name: str
    relative_path: str
    permalink: str
    title: str
    note_type: str
    content: str
    tags: list[str]
    project: str | None
    created_at: datetime
    updated_at: datetime
    supersedes: int | None = None
    superseded_by: int | None = None

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Note:
        """Create a Note from API response data."""
        return cls(
            id=data["id"],
            vault_name=data["vault_name"],
            relative_path=data["relative_path"],
            permalink=data["permalink"],
            title=data["title"],
            note_type=data["note_type"],
            content=data.get("content", ""),
            tags=data.get("tags", []),
            project=data.get("project"),
            created_at=datetime.fromisoformat(data["created_at"].replace("Z", "+00:00")),
            updated_at=datetime.fromisoformat(data["updated_at"].replace("Z", "+00:00")),
            supersedes=data.get("supersedes"),
            superseded_by=data.get("superseded_by"),
        )


@dataclass
class Project:
    """Represents a project."""

    name: str
    note_count: int

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Project:
        """Create a Project from API response data."""
        return cls(name=data["name"], note_count=data["note_count"])


class ObsidianMemoryError(Exception):
    """Base exception for API errors."""

    def __init__(self, message: str, status_code: int | None = None):
        super().__init__(message)
        self.status_code = status_code


class NotFoundError(ObsidianMemoryError):
    """Resource not found."""

    pass


class AuthenticationError(ObsidianMemoryError):
    """Authentication failed."""

    pass


class RateLimitError(ObsidianMemoryError):
    """Rate limit exceeded."""

    def __init__(self, message: str, retry_after: int):
        super().__init__(message, 429)
        self.retry_after = retry_after


class ObsidianMemoryClient:
    """Client for the Obsidian-Memory API."""

    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        auth_token: str | None = None,
        timeout: float = 30.0,
    ):
        """Initialize the client.

        Args:
            base_url: API base URL
            auth_token: Bearer token for authentication (optional)
            timeout: Request timeout in seconds
        """
        self.base_url = base_url.rstrip("/")
        self.auth_token = auth_token
        self.timeout = timeout
        self._client = httpx.Client(timeout=timeout)

    def _headers(self) -> dict[str, str]:
        """Get request headers."""
        headers = {"Content-Type": "application/json"}
        if self.auth_token:
            headers["Authorization"] = f"Bearer {self.auth_token}"
        return headers

    def _handle_response(self, response: httpx.Response) -> dict[str, Any]:
        """Handle API response, raising appropriate errors."""
        if response.status_code == 401:
            raise AuthenticationError("Authentication required", 401)
        if response.status_code == 403:
            raise AuthenticationError("Invalid credentials", 403)
        if response.status_code == 404:
            raise NotFoundError("Resource not found", 404)
        if response.status_code == 429:
            retry_after = int(response.headers.get("Retry-After", 60))
            raise RateLimitError("Rate limit exceeded", retry_after)

        response.raise_for_status()
        return response.json()

    def _get(self, path: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
        """Make a GET request."""
        url = f"{self.base_url}{path}"
        response = self._client.get(url, headers=self._headers(), params=params)
        return self._handle_response(response)

    def _post(self, path: str, data: dict[str, Any] | None = None) -> dict[str, Any]:
        """Make a POST request."""
        url = f"{self.base_url}{path}"
        response = self._client.post(url, headers=self._headers(), json=data)
        return self._handle_response(response)

    def _put(self, path: str, data: dict[str, Any] | None = None) -> dict[str, Any]:
        """Make a PUT request."""
        url = f"{self.base_url}{path}"
        response = self._client.put(url, headers=self._headers(), json=data)
        return self._handle_response(response)

    def _delete(self, path: str) -> dict[str, Any]:
        """Make a DELETE request."""
        url = f"{self.base_url}{path}"
        response = self._client.delete(url, headers=self._headers())
        return self._handle_response(response)

    # Health & Metrics

    def health(self) -> dict[str, Any]:
        """Check API health status."""
        return self._get("/health")

    def metrics(self) -> dict[str, Any]:
        """Get system metrics."""
        return self._get("/metrics")

    # Notes

    def list_notes(
        self,
        vault: str | None = None,
        project: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> tuple[list[Note], int]:
        """List notes with optional filtering.

        Args:
            vault: Filter by vault name
            project: Filter by project
            limit: Maximum results to return
            offset: Pagination offset

        Returns:
            Tuple of (notes list, total count)
        """
        params = {"limit": limit, "offset": offset}
        if vault:
            params["vault"] = vault
        if project:
            params["project"] = project

        data = self._get("/api/notes", params)
        notes = [Note.from_dict(n) for n in data["notes"]]
        return notes, data["total"]

    def get_note(self, note_id: int) -> Note:
        """Get a note by ID.

        Args:
            note_id: Note ID

        Returns:
            Note object

        Raises:
            NotFoundError: If note doesn't exist
        """
        data = self._get(f"/api/notes/{note_id}")
        return Note.from_dict(data)

    def create_note(
        self,
        vault_name: str,
        title: str,
        content: str,
        relative_path: str | None = None,
        note_type: str = "note",
        project: str | None = None,
        tags: list[str] | None = None,
    ) -> Note:
        """Create a new note.

        Args:
            vault_name: Vault to create note in
            title: Note title
            content: Note content (Markdown)
            relative_path: Path within vault (auto-generated if not provided)
            note_type: Type of note (note, decision, learning, etc.)
            project: Project name
            tags: List of tags

        Returns:
            Created note object
        """
        if relative_path is None:
            # Generate path from title
            safe_title = title.lower().replace(" ", "-")
            relative_path = f"notes/{safe_title}.md"

        data = self._post(
            "/api/notes",
            {
                "vault_name": vault_name,
                "relative_path": relative_path,
                "title": title,
                "content": content,
                "note_type": note_type,
                "project": project,
                "tags": tags or [],
            },
        )
        return Note.from_dict(data)

    def update_note(
        self,
        note_id: int,
        title: str | None = None,
        content: str | None = None,
        tags: list[str] | None = None,
        project: str | None = None,
    ) -> Note:
        """Update a note.

        Args:
            note_id: Note ID to update
            title: New title (optional)
            content: New content (optional)
            tags: New tags (optional)
            project: New project (optional)

        Returns:
            Updated note object
        """
        updates = {}
        if title is not None:
            updates["title"] = title
        if content is not None:
            updates["content"] = content
        if tags is not None:
            updates["tags"] = tags
        if project is not None:
            updates["project"] = project

        data = self._put(f"/api/notes/{note_id}", updates)
        return Note.from_dict(data)

    def delete_note(self, note_id: int) -> None:
        """Delete a note.

        Args:
            note_id: Note ID to delete
        """
        self._delete(f"/api/notes/{note_id}")

    def search(
        self,
        query: str,
        vault: str | None = None,
        project: str | None = None,
        note_type: str | None = None,
        tags: list[str] | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> tuple[list[Note], int]:
        """Search notes with full-text search.

        Args:
            query: Search query
            vault: Filter by vault
            project: Filter by project
            note_type: Filter by note type
            tags: Filter by tags (all must match)
            limit: Maximum results
            offset: Pagination offset

        Returns:
            Tuple of (matching notes, total count)
        """
        search_params = {
            "query": query,
            "limit": limit,
            "offset": offset,
        }
        if vault:
            search_params["vault"] = vault
        if project:
            search_params["project"] = project
        if note_type:
            search_params["note_type"] = note_type
        if tags:
            search_params["tags"] = tags

        data = self._post("/api/notes/search", search_params)
        notes = [Note.from_dict(n) for n in data["notes"]]
        return notes, data["total"]

    def supersede_note(
        self,
        old_note_id: int,
        new_note_id: int,
        reason: str | None = None,
    ) -> dict[str, Any]:
        """Mark a note as superseded by another note.

        Creates bi-directional relationships for knowledge evolution tracking.

        Args:
            old_note_id: ID of the note being replaced
            new_note_id: ID of the replacement note
            reason: Optional reason for supersession

        Returns:
            Dictionary with operation result
        """
        return self._post(
            "/api/notes/supersede",
            {
                "old_note_id": old_note_id,
                "new_note_id": new_note_id,
                "reason": reason,
            },
        )

    # Projects

    def list_projects(self) -> list[Project]:
        """List all projects with note counts.

        Returns:
            List of projects
        """
        data = self._get("/api/projects")
        return [Project.from_dict(p) for p in data["projects"]]

    def get_project_notes(
        self,
        project_name: str,
        limit: int = 50,
        offset: int = 0,
    ) -> tuple[list[dict[str, Any]], int]:
        """Get notes for a specific project.

        Args:
            project_name: Project name
            limit: Maximum results
            offset: Pagination offset

        Returns:
            Tuple of (note summaries, total count)
        """
        data = self._get(
            f"/api/projects/{project_name}/notes",
            params={"limit": limit, "offset": offset},
        )
        return data["notes"], data["total_count"]

    # Graph

    def get_graph(self) -> dict[str, Any]:
        """Get the full knowledge graph.

        Returns:
            Dictionary with 'nodes' and 'edges'
        """
        return self._get("/api/graph")

    def get_neighbors(self, node_id: int) -> dict[str, Any]:
        """Get neighbors for a graph node.

        Args:
            node_id: Node ID

        Returns:
            Dictionary with neighbor information
        """
        return self._get(f"/api/graph/nodes/{node_id}/neighbors")

    def close(self) -> None:
        """Close the HTTP client."""
        self._client.close()

    def __enter__(self) -> ObsidianMemoryClient:
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()


# Example usage
if __name__ == "__main__":
    # Basic usage example
    with ObsidianMemoryClient(base_url="http://localhost:8000") as client:
        # Check health
        health = client.health()
        print(f"API Status: {health['status']}")

        # List notes
        notes, total = client.list_notes(limit=5)
        print(f"\nFound {total} notes:")
        for note in notes:
            print(f"  - {note.title} ({note.vault_name})")

        # Search example
        results, count = client.search("python", limit=3)
        print(f"\nSearch results for 'python' ({count} total):")
        for note in results:
            print(f"  - {note.title}")

        # Create a note example (commented out to avoid side effects)
        # note = client.create_note(
        #     vault_name="my-vault",
        #     title="API Test Note",
        #     content="# Test Note\n\nCreated via Python SDK.",
        #     tags=["test", "api"],
        # )
        # print(f"\nCreated note: {note.id} - {note.title}")

        # Get graph stats
        graph = client.get_graph()
        print(f"\nKnowledge graph: {len(graph['nodes'])} nodes, {len(graph['edges'])} edges")

File: examples/typescript/obsidian-memory-client.ts
/**
 * Obsidian-Memory TypeScript SDK Example
 *
 * A simple client library for interacting with the Obsidian-Memory API.
 *
 * Usage:
 *   import { ObsidianMemoryClient } from './obsidian-memory-client';
 *
 *   const client = new ObsidianMemoryClient({
 *     baseUrl: 'http://localhost:8000',
 *     authToken: 'your-token',  // Optional if auth is disabled
 *   });
 *
 *   // List notes
 *   const { notes, total } = await client.listNotes({ limit: 10 });
 *
 *   // Search notes
 *   const results = await client.search('machine learning', { project: 'research' });
 *
 *   // Create a note
 *   const note = await client.createNote({
 *     vaultName: 'my-vault',
 *     title: 'New Note',
 *     content: '# My Note\n\nContent here...',
 *   });
 */

// Types

export interface Note {
  id: number;
  vault_name: string;
  relative_path: string;
  permalink: string;
  title: string;
  note_type: string;
  content: string;
  tags: string[];
  project: string | null;
  created_at: string;
  updated_at: string;
  supersedes?: number | null;
  superseded_by?: number | null;
}

export interface Project {
  name: string;
  note_count: number;
}

export interface GraphNode {
  id: number;
  title: string;
  permalink: string;
  vault_name: string;
  note_type: string;
  project: string | null;
  tags: string[];
}

export interface GraphEdge {
  source: number;
  target: number;
  target_title: string;
  type: string;
  context: string | null;
  weight: number;
}

export interface Graph {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface HealthStatus {
  status: string;
  version?: string;
  vault_connected?: boolean;
  timestamp?: string;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  limit: number;
  offset: number;
}

export interface SupersedeResult {
  old_note_id: number;
  new_note_id: number;
  old_note_title: string;
  new_note_title: string;
  message: string;
}

// Errors

export class ObsidianMemoryError extends Error {
  constructor(
    message: string,
    public statusCode?: number
  ) {
    super(message);
    this.name = 'ObsidianMemoryError';
  }
}

export class NotFoundError extends ObsidianMemoryError {
  constructor(message: string = 'Resource not found') {
    super(message, 404);
    this.name = 'NotFoundError';
  }
}

export class AuthenticationError extends ObsidianMemoryError {
  constructor(message: string = 'Authentication failed') {
    super(message, 401);
    this.name = 'AuthenticationError';
  }
}

export class RateLimitError extends ObsidianMemoryError {
  constructor(
    message: string = 'Rate limit exceeded',
    public retryAfter: number = 60
  ) {
    super(message, 429);
    this.name = 'RateLimitError';
  }
}

// Client Options

export interface ClientOptions {
  baseUrl?: string;
  authToken?: string;
  timeout?: number;
}

// Request Options

export interface ListNotesOptions {
  vault?: string;
  project?: string;
  limit?: number;
  offset?: number;
}

export interface SearchOptions {
  vault?: string;
  project?: string;
  noteType?: string;
  tags?: string[];
  limit?: number;
  offset?: number;
}

export interface CreateNoteOptions {
  vaultName: string;
  title: string;
  content: string;
  relativePath?: string;
  noteType?: string;
  project?: string;
  tags?: string[];
}

export interface UpdateNoteOptions {
  title?: string;
  content?: string;
  tags?: string[];
  project?: string;
}

// Client Implementation

export class ObsidianMemoryClient {
  private baseUrl: string;
  private authToken?: string;
  private timeout: number;

  constructor(options: ClientOptions = {}) {
    this.baseUrl = (options.baseUrl || 'http://localhost:8000').replace(
      /\/$/,
      ''
    );
    this.authToken = options.authToken;
    this.timeout = options.timeout || 30000;
  }

  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    if (this.authToken) {
      headers['Authorization'] = `Bearer ${this.authToken}`;
    }
    return headers;
  }

  private async handleResponse<T>(response: Response): Promise<T> {
    if (response.status === 401) {
      throw new AuthenticationError('Authentication required');
    }
    if (response.status === 403) {
      throw new AuthenticationError('Invalid credentials');
    }
    if (response.status === 404) {
      throw new NotFoundError();
    }
    if (response.status === 429) {
      const retryAfter = parseInt(response.headers.get('Retry-After') || '60');
      throw new RateLimitError('Rate limit exceeded', retryAfter);
    }

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new ObsidianMemoryError(
        error.detail || `HTTP ${response.status}`,
        response.status
      );
    }

    return response.json();
  }

  private async get<T>(
    path: string,
    params?: Record<string, string | number>
  ): Promise<T> {
    const url = new URL(`${this.baseUrl}${path}`);
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          url.searchParams.append(key, String(value));
        }
      });
    }

    const response = await fetch(url.toString(), {
      method: 'GET',
      headers: this.getHeaders(),
      signal: AbortSignal.timeout(this.timeout),
    });

    return this.handleResponse<T>(response);
  }

  private async post<T>(path: string, data?: unknown): Promise<T> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: data ? JSON.stringify(data) : undefined,
      signal: AbortSignal.timeout(this.timeout),
    });

    return this.handleResponse<T>(response);
  }

  private async put<T>(path: string, data?: unknown): Promise<T> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'PUT',
      headers: this.getHeaders(),
      body: data ? JSON.stringify(data) : undefined,
      signal: AbortSignal.timeout(this.timeout),
    });

    return this.handleResponse<T>(response);
  }

  private async delete<T>(path: string): Promise<T> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'DELETE',
      headers: this.getHeaders(),
      signal: AbortSignal.timeout(this.timeout),
    });

    return this.handleResponse<T>(response);
  }

  // Health & Metrics

  async health(): Promise<HealthStatus> {
    return this.get<HealthStatus>('/health');
  }

  async metrics(): Promise<Record<string, unknown>> {
    return this.get('/metrics');
  }

  // Notes

  async listNotes(
    options: ListNotesOptions = {}
  ): Promise<{ notes: Note[]; total: number }> {
    const params: Record<string, string | number> = {
      limit: options.limit || 50,
      offset: options.offset || 0,
    };
    if (options.vault) params.vault = options.vault;
    if (options.project) params.project = options.project;

    const data = await this.get<{ notes: Note[]; total: number }>(
      '/api/notes',
      params
    );
    return data;
  }

  async getNote(noteId: number): Promise<Note> {
    return this.get<Note>(`/api/notes/${noteId}`);
  }

  async createNote(options: CreateNoteOptions): Promise<Note> {
    const relativePath =
      options.relativePath ||
      `notes/${options.title.toLowerCase().replace(/\s+/g, '-')}.md`;

    return this.post<Note>('/api/notes', {
      vault_name: options.vaultName,
      relative_path: relativePath,
      title: options.title,
      content: options.content,
      note_type: options.noteType || 'note',
      project: options.project,
      tags: options.tags || [],
    });
  }

  async updateNote(noteId: number, options: UpdateNoteOptions): Promise<Note> {
    const updates: Record<string, unknown> = {};
    if (options.title !== undefined) updates.title = options.title;
    if (options.content !== undefined) updates.content = options.content;
    if (options.tags !== undefined) updates.tags = options.tags;
    if (options.project !== undefined) updates.project = options.project;

    return this.put<Note>(`/api/notes/${noteId}`, updates);
  }

  async deleteNote(noteId: number): Promise<void> {
    await this.delete(`/api/notes/${noteId}`);
  }

  async search(
    query: string,
    options: SearchOptions = {}
  ): Promise<{ notes: Note[]; total: number }> {
    const searchParams: Record<string, unknown> = {
      query,
      limit: options.limit || 50,
      offset: options.offset || 0,
    };
    if (options.vault) searchParams.vault = options.vault;
    if (options.project) searchParams.project = options.project;
    if (options.noteType) searchParams.note_type = options.noteType;
    if (options.tags) searchParams.tags = options.tags;

    return this.post<{ notes: Note[]; total: number }>(
      '/api/notes/search',
      searchParams
    );
  }

  async supersedeNote(
    oldNoteId: number,
    newNoteId: number,
    reason?: string
  ): Promise<SupersedeResult> {
    return this.post<SupersedeResult>('/api/notes/supersede', {
      old_note_id: oldNoteId,
      new_note_id: newNoteId,
      reason,
    });
  }

  // Projects

  async listProjects(): Promise<Project[]> {
    const data = await this.get<{ projects: Project[] }>('/api/projects');
    return data.projects;
  }

  async getProjectNotes(
    projectName: string,
    options: { limit?: number; offset?: number } = {}
  ): Promise<{ notes: Array<Record<string, unknown>>; total: number }> {
    const data = await this.get<{
      notes: Array<Record<string, unknown>>;
      total_count: number;
    }>(`/api/projects/${encodeURIComponent(projectName)}/notes`, {
      limit: options.limit || 50,
      offset: options.offset || 0,
    });
    return { notes: data.notes, total: data.total_count };
  }

  // Graph

  async getGraph(): Promise<Graph> {
    return this.get<Graph>('/api/graph');
  }

  async getNeighbors(
    nodeId: number
  ): Promise<{ incoming: GraphEdge[]; outgoing: GraphEdge[] }> {
    return this.get(`/api/graph/nodes/${nodeId}/neighbors`);
  }
}

// Example usage
async function main() {
  const client = new ObsidianMemoryClient({
    baseUrl: 'http://localhost:8000',
    // authToken: 'your-token-here',  // Uncomment if auth is enabled
  });

  try {
    // Check health
    const health = await client.health();
    console.log(`API Status: ${health.status}`);

    // List notes
    const { notes, total } = await client.listNotes({ limit: 5 });
    console.log(`\nFound ${total} notes:`);
    notes.forEach((note) => {
      console.log(`  - ${note.title} (${note.vault_name})`);
    });

    // Search example
    const searchResults = await client.search('python', { limit: 3 });
    console.log(
      `\nSearch results for 'python' (${searchResults.total} total):`
    );
    searchResults.notes.forEach((note) => {
      console.log(`  - ${note.title}`);
    });

    // Create a note example (commented out to avoid side effects)
    // const newNote = await client.createNote({
    //   vaultName: 'my-vault',
    //   title: 'API Test Note',
    //   content: '# Test Note\n\nCreated via TypeScript SDK.',
    //   tags: ['test', 'api'],
    // });
    // console.log(`\nCreated note: ${newNote.id} - ${newNote.title}`);

    // Get graph stats
    const graph = await client.getGraph();
    console.log(
      `\nKnowledge graph: ${graph.nodes.length} nodes, ${graph.edges.length} edges`
    );
  } catch (error) {
    if (error instanceof RateLimitError) {
      console.error(`Rate limited. Retry after ${error.retryAfter} seconds.`);
    } else if (error instanceof AuthenticationError) {
      console.error('Authentication failed:', error.message);
    } else if (error instanceof NotFoundError) {
      console.error('Resource not found');
    } else {
      console.error('Error:', error);
    }
  }
}

// Run example if executed directly
main().catch(console.error);
